### PR TITLE
rustc: Make `def_kind` mandatory for all `DefId`s

### PR DESCRIPTION
**Pull request by petrochenkov**
Prerequisite for https://github.com/rust-lang/rust/pull/118188.

### CODE REVIEW

File: compiler/rustc_metadata/src/rmeta/mod.rs
Comment (compiler-errors) [inline compiler/rustc_metadata/src/rmeta/mod.rs:389]: Why is it preferable to put it into defaulted set?

Could we just unwrap the option and keep it in the optional set, or do they behave differently for perf?

File: compiler/rustc_middle/src/hir/map/mod.rs
Comment (compiler-errors) [inline compiler/rustc_middle/src/hir/map/mod.rs:894]: What codepath triggers this specifically? Just curious.

File: compiler/rustc_metadata/src/rmeta/encoder.rs
Comment (compiler-errors) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:1399]: We could just move the def-id encode up to the top then early continue?

**Review by compiler-errors — COMMENTED** [review]
👍 like this a lot. a few comments (and i know perf is running)

File: compiler/rustc_metadata/src/rmeta/mod.rs
Comment (petrochenkov) [reply compiler/rustc_metadata/src/rmeta/mod.rs:389]: I just didn't like having an `Option` in the table when the def kind is not actually optional, there should be no perf difference here.
But it may be marginally cleaner (and less changes) to keep the `Option`, yes.

File: compiler/rustc_middle/src/hir/map/mod.rs
Comment (petrochenkov) [reply compiler/rustc_middle/src/hir/map/mod.rs:894]: `def_span` for anonymous constants without HIR.

File: compiler/rustc_metadata/src/rmeta/encoder.rs
Comment (petrochenkov) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:1399]: I suspect that the skipped encoding for a subset of anon consts is a source of bugs that we do not observe only because uses of `#[rustc_legacy_const_generics]` are very rare.
So we should rather try to encode `type_of` and `generics_of` (the only data that is skipped now), than to go back to skipping everything.

Comment (compiler-errors) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:1399]: Makes sense. Can you add a FIXME for those HIR-less queries?

File: compiler/rustc_metadata/src/rmeta/mod.rs
Comment (compiler-errors) [reply compiler/rustc_metadata/src/rmeta/mod.rs:389]: I think I would prefer the optional one, though I guess not particularly strongly, since it means we can go without the `Default` impl, for example.

File: compiler/rustc_metadata/src/rmeta/encoder.rs
Comment (petrochenkov) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:1399]: There is a FIXME, in the comment above.

File: compiler/rustc_metadata/src/rmeta/encoder.rs
Comment (cjgillot) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:1395]: Should we give them a dummy hir node to avoid those ices?

File: compiler/rustc_middle/src/hir/map/mod.rs
Comment (cjgillot) [reply compiler/rustc_middle/src/hir/map/mod.rs:894]: Should we keep a FIXME here too?

File: compiler/rustc_metadata/src/rmeta/encoder.rs
Comment (petrochenkov) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:1395]: Maybe, I just didn't want to get into that in this PR, https://github.com/rust-lang/rust/pull/118188 is not blocked on it.

I tried to look what exactly happens with `rustc_legacy_const_generics`, but it's still not clear to me.
There seems to be some mismatch between what def collector and AST lowering do.

File: compiler/rustc_middle/src/hir/map/mod.rs
Comment (petrochenkov) [reply compiler/rustc_middle/src/hir/map/mod.rs:894]: Not all passed `HirId`s are required have HIR nodes or `Ident`s, so the new behavior looks reasonable to me, we just didn't hit this case before.

File: compiler/rustc_metadata/src/rmeta/mod.rs
Comment (petrochenkov) [reply compiler/rustc_metadata/src/rmeta/mod.rs]: Reverted this change.

**Review by compiler-errors — APPROVED** [review]
r=me unless you want another round of review from cjgillot

**Comment by petrochenkov** [comment]
@bors r=compiler-errors
I'd want to see https://github.com/rust-lang/rust/pull/118250#discussion_r1404702779 addressed somehow, just don't have time to investigate right now.

**Comment by nnethercote** [comment]
If anything, this looks like a tiny performance win.

@rustbot label: +perf-regression-triaged

**Pull request merged**
